src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,145 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+    
     @Override
     public void add(T value) {
+        Node<T> newNode;
+        if (size == 0) {   
+            newNode = new Node<>(null, value, null); 
+            head = newNode;
+            tail = newNode;
+        } else {
+            newNode = new Node<>(tail, value, null); 
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> newNode;
+            if (index == 0) {
+                newNode = new Node<>(null, value, head);
+                head.prev = newNode;
+                head = newNode;
+            } else {
+                checkIndexValidity(index);
+                Node<T> currentNode = getNodeByIndex(index);
+                newNode = new Node<>(currentNode.prev, value, currentNode);
+                currentNode.prev.next = newNode;
+                currentNode.prev = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexValidity(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexValidity(index);
+        Node<T> currentNode = getNodeByIndex(index);
+        T prevValue = currentNode.value;
+        currentNode.value = value;
+        return prevValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexValidity(index);
+        Node<T> currentNode = getNodeByIndex(index);
+        T removedValue = currentNode.value;
+        if (size == 1) {
+            head = null;
+            tail = null;
+        } else if (index == 0) {
+            currentNode.next.prev = null;
+            head = currentNode.next;
+        } else if (index == size - 1) {
+            currentNode.prev.next = null;
+            tail = currentNode.prev;
+        } else {
+            currentNode.prev.next = currentNode.next;
+            currentNode.next.prev = currentNode.prev;
+        }
+        size--;
+        return removedValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (currentNode.value == object 
+                    || (currentNode.value != null && currentNode.value.equals(object))) {
+                remove(i);
```

Comment:
you can call unlink(currentNode) here - to avoid iteration over the list in remove(index) method